### PR TITLE
Running constraints of different timings

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Signed-off-by: Francesco Guardiani <francescoguard@gmail.com>

# Changes

- :broom: Now `Setup`, `Teardown` and `Requirement` run ordered and sequentially, while `Assert` is tagged to run in parallel. This makes the setup, teardown and requirement phases of the test predictable
- :gift: Test for the runtime constraints of the tests

### CODE REVIEW

File: pkg/feature/collapse.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+// Steps with Assert timing are run in parallel
+func CollapseSteps(steps []Step) []Step {
+	var setup *Step
+	var requirement *Step
+	var asserts []Step
+	var teardown *Step
+
+	for i, s := range steps {
+		if s.T == Setup {
+			setup = composeStep(setup, &steps[i])
```
Comment (n3wscott) [inline pkg/feature/collapse.go:19]: why not the way it was?

I think this is much more confusing in the output of the test:

```
=== test1 AND test2 AND test3
```

if that fails, how are you going to debug which is having the problem?

Comment (slinkydeveloper) [reply pkg/feature/collapse.go:19]: That's fair, but I think we can solve it somehow in a followup, for example we could manually print these:

```
==== test1
==== test2
==== test3
```

Comment (slinkydeveloper) [reply pkg/feature/collapse.go:19]: The important thing here is that we force the ordered and sequential execution

File: pkg/feature/collapse.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+)
+
+// CollapseSteps is used by the framework to impose the execution constraints of the different steps
+// Steps with Setup, Requirement and Teardown timings are run sequentially in order
+// Steps with Assert timing are run in parallel
+func CollapseSteps(steps []Step) []Step {
+	var setup *Step
+	var requirement *Step
+	var asserts []Step
+	var teardown *Step
```
Comment (vagababov) [inline pkg/feature/collapse.go:15]:
```suggestion
	var (
	     setup, requirement, teardown *Step
	     asserts []Step
	)
```

File: pkg/feature/collapse.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+	for i, s := range steps {
+		if s.T == Setup {
```
Comment (vagababov) [inline pkg/feature/collapse.go:18]: switch here

**Pull request merged**